java-checks/src/main/java/org/sonar/java/checks/TypeParametersShadowingCheck.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.TypeParameters;
+
+@Rule(key = "S4977")
+public class TypeParametersShadowingCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private static final String ISSUE_MESSAGE = "Rename \"%s\" which hides a type parameter from the outer scope.";
+
+  private JavaFileScannerContext context;
+
+  private Map<String, IdentifierTree> currentTypeParametersInScope;
+  private Deque<Map<String, IdentifierTree>> scopes;

```

Comment:
You don't need an explicit stack when using `BaseTreeVisitor`. You can just set `currentTypeParametersInScope` at the beginning of `processTree` (storing the old value in a local variable) and set it back to its old value at the end.

Reply:
Indeed, nice improvment!